File: assign13/q6.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jun 15 14:18:31 2023

@author: harsha
"""

class node:
    def __init__(self,data):
        self.data = data
        self.next = None
        
class ll:
    def __init__(self):
        self.head = None
        
def insert(root, item):
    temp = node(item)
      
    if (root == None):
        root = temp
    else :
        ptr = root
        while (ptr.next != None):
            ptr = ptr.next
        ptr.next = temp
      
    return root
        
def arraytoll(arr):
    root = None
    for i in range(len(arr)):
        root = insert(root, arr[i])
    return root

def display(root):
    while (root != None) :
        print(root.data, end = "->")
        root = root.next


def merge(h1,h2):
    if(h1==None):
        return h2
    if(h2==None):
        return h1
    if(h1.data<h2.data):
        h1.next = merge(h1.next,h2)
        return h1
    else:
        h2.next = merge(h1,h2.next)
        return h2


    

# link = ll()
# link.head = node(1)
arr1 = [5,10,15]
arr2 = [2,3,20]
root1 = arraytoll(arr1)
root2 = arraytoll(arr2)
ans = merge(root1,root2)
display(ans)

File: assign14/q6.cpp
#include <iostream>
#include <bits/stdc++.h>


using namespace std;

struct node{
	int data;
	node* next;
	//constrctor
	node(int x){
		this->data = x;
		this->next =NULL;
	}
};

node* array2ll(vector<int> &arr){
	node* head = new node(arr[0]);
	node* temp = head;
	for(int i=1;i<arr.size();i++){
		node* n = new node(arr[i]);
		temp->next = n;
		temp = temp->next;
	}
	return head;
}


void display(node *head){
	node *temp=head;
	while(temp!=NULL){
		cout<<temp->data<<"->";
		temp = temp->next;
	}
	cout<<endl;
}

//////////////  ANSWER /////////////////
//assuming shift is less than length of ll
node* shift(node* head,int k) {
    node* cur = head;
    for(int i=1;i<k;i++){
        cur = cur->next;
    }
    node* ans = cur->next,*temp = cur->next;
    cur->next = NULL;
    while(temp->next) temp = temp->next;
    temp->next = head;
    return ans;

}

int main(){
	vector<int> v1{1,2,3,4,5,6,7,8};
	node* head = array2ll(v1);
	display(head); 
    node* ans = shift(head,4);
	display(ans); 

	return 0;
}